### ISSUE TITLE
MachinePool version is optional but implicitly required by AzureManagedMachinePool

### ISSUE THREAD

**Issue by mjnovice** [kubernetes-sigs/cluster-api-provider-azure#3435]
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
Provision an AKS managed cluster, with the following example manifest

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: sfd-${NAME}
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: ${NAME}
  namespace: sfd-${NAME}
spec:
  clusterNetwork:
    services:
      cidrBlocks:
      - 192.168.0.0/16
  controlPlaneRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureManagedControlPlane
    name: ${NAME}
    namespace: sfd-${NAME}
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureManagedCluster
    name: ${NAME}
    namespace: sfd-${NAME}
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedControlPlane
metadata:
  labels:
    cluster.x-k8s.io/cluster-name: ${NAME}
  name: ${NAME}
  namespace: sfd-${NAME}
spec:
  subscriptionID: 64fdac10-935b-40e6-bf28-f7dc093f7f76
  additionalTags:
    Owner: mayank.jha@uipath.com
    Project: Service Fabric
  apiServerAccessProfile:
    enablePrivateCluster: true
  identityRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureClusterIdentity
    name: ${NAME}
  loadBalancerSKU: Standard
  location: westeurope
  networkPlugin: azure
  resourceGroupName: ${NAME}
  version: v1.23.12
  virtualNetwork:
    cidrBlock: 10.0.0.0/8
    name: ${NAME}
    resourceGroup: ${NAME}
    subnet:
      cidrBlock: 10.240.0.0/16
      name: ${NAME}
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedMachinePool
metadata:
  name: ${NAME}-0
  namespace: sfd-${NAME}
spec:
  enableNodePublicIP: false
  maxPods: 100
  mode: System
  name: pool0
  nodeLabels:
    dummykey: dummyval
  osDiskSizeGB: 512
  osDiskType: Managed
  osType: Linux
  scaling:
    maxSize: 8
    minSize: 3
  sku: Standard_D8s_v3
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: MachinePool
metadata:
  name: ${NAME}-0
  namespace: sfd-${NAME}
spec:
  clusterName: ${NAME}
  replicas: 3
  template:
    metadata: {}
    spec:
      bootstrap:
        dataSecretName: ""
      clusterName: ${NAME}
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: AzureManagedMachinePool
        name: ${NAME}-0
        namespace: sfd-${NAME}
      version: v1.23.12
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureClusterIdentity
metadata:
  name: ${NAME}
  namespace: sfd-${NAME}
spec:
  allowedNamespaces:
    list:
    - sfd-${NAME}
  clientID: 0ddc4f44-26cc-4018-9336-83e375627d99
  clientSecret:
    name: azcapi-infrasfea
    namespace: as-air-cluster-system
  tenantID: d8353d2a-b153-4d17-8827-902c51f72357
  type: ServicePrincipal
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedCluster
metadata:
  name: ${NAME}
  namespace: sfd-${NAME}
```

The machinePool seems to be in a loop of `update` even when the spec doesn't change at all. 

```
 conditions:
  - lastTransitionTime: "2023-04-17T00:43:31Z"
    message: 'agentpools failed to create or update. err: failed to get desired parameters
      for resource devmj-mj-test5/pool0 (service: agentpools): Unable to update existing
      agent pool in non terminal state. Agent pool must be in one of the following
      provisioning states: Canceled, Failed, or Succeeded. Actual state: Updating.
      Object will be requeued after 20s'
    reason: Failed
    severity: Error
    status: "False"
    type: Ready
```

<img width="915" alt="image" src="https://user-images.githubusercontent.com/3083824/232353918-572e5d21-58f8-45f4-b15c-ac72de038cc2.png">

On the Azure portal, there seems to be redundant set of updates.

**What did you expect to happen:**

There shouldn't be update calls when there is no apparent update to the `MachinePool`

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: `1.8.2`
- Kubernetes version: (use `kubectl version`): `Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-11-02T03:24:50Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/arm64"}`
- OS (e.g. from `/etc/os-release`): macOS 13.3.1 (22E261)

**Comment by nojnhuh** [comment]
This smells like #3009...

@mjnovice Can you try bumping the log verbosity of the capz-controller-manager Pod to at least `-v=4` and checking the logs for any lines that include `found a diff` and share what you find? That should point us to exactly what CAPZ is trying to update.

**Comment by jamesgoodhouse** [comment]
It's possible this is the same issue I reported in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3413 and have opened a PR to fix (https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3414).

**Comment by mjnovice** [comment]
@nojnhuh I didn't get any logs even after bumping the verbosity, and this is hard to reproduce consistently on subsequent runs.

**Comment by nojnhuh** [comment]
@mjnovice So `kubectl logs -n capz-system deploy/capz-controller-manager` succeeds doesn't show anything?

**Comment by mjnovice** [comment]
@nojnhuh nope.

**Comment by mjnovice** [comment]
This is happening consistently when Azure private cluster option is enabled.

**Comment by mjnovice** [comment]
Okay, I might have overlooked it.

```
I0627 23:21:05.273226       1 spec.go:247] agentpools.Service.Parameters "msg"="found a diff between the desired spec and the existing agentpool" "AzureManagedMachinePool"={"name":"ci-mj3-0","namespace":"sfd-ci-mj3"} "controller"="azuremanagedmachinepool" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedMachinePool" "difference"="  containerservice.AgentPool{\n  \tResponse: {},\n  \tManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{\n  \t\t... // 15 identical fields\n  \t\tType:                \"\",\n  \t\tMode:                \"System\",\n- \t\tOrchestratorVersion: nil,\n+ \t\tOrchestratorVersion: &\"1.25.6\",\n  \t\tNodeImageVersion:    nil,\n  \t\tUpgradeSettings:     nil,\n  \t\t... // 19 identical fields\n  \t},\n  \tID:   nil,\n  \tName: nil,\n  \tType: nil,\n  }\n" "name"="ci-mj3-0" "namespace"="sfd-ci-mj3" "reconcileID"="f727cf64-b700-43a2-838d-6166dd5b187c" "x-ms-correlation-request-id"="949e923e-636e-466c-9e18-16fc59401463"
I0627 23:21:18.658765       1 spec.go:427] managedclusters.Service.Parameters "msg"="found a diff between the desired spec and the existing managed cluster" "AzureManagedControlPlane"={"name":"ci-mj3","namespace":"sfd-ci-mj3"} "controller"="azuremanagedcontrolplane" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedControlPlane" "difference"="  &containerservice.ManagedCluster{\n  \t... // 2 identical fields\n  \tExtendedLocation: nil,\n  \tIdentity:         nil,\n  \tManagedClusterProperties: &containerservice.ManagedClusterProperties{\n  \t\t... // 20 identical fields\n  \t\tAutoUpgradeProfile: nil,\n  \t\tAutoScalerProfile:  nil,\n  \t\tAPIServerAccessProfile: &containerservice.ManagedClusterAPIServerAccessProfile{\n- \t\t\tAuthorizedIPRanges:   &nil,\n+ \t\t\tAuthorizedIPRanges:   nil,\n  \t\t\tEnablePrivateCluster: nil,\n  \t\t\tPrivateDNSZone:       nil,\n  \t\t\t... // 2 identical fields\n  \t\t},\n  \t\tDiskEncryptionSetID: nil,\n  \t\tIdentityProfile:     nil,\n  \t\t... // 5 identical fields\n  \t},\n  \tTags:     nil,\n  \tLocation: nil,\n  \t... // 4 identical fields\n  }\n" "name"="ci-mj3" "namespace"="sfd-ci-mj3" "reconcileID"="bc24f73d-f142-4600-9eb3-86f8fc364edf" "x-ms-correlation-request-id"="ad75d889-ad36-450b-8cfd-6e768073995d"
```

`AzureManagedMachinePool` seems to be reconciling with an apparent diff for `OrchestratorVersion` 

Perhaps that is not used when we provision a private AKS cluster ? cc - @nojnhuh

**Comment by nojnhuh** [comment]
Are you sure `spec.template.spec.version` is defined on the MachinePool? CAPZ seems to assume that's specified, and with a fairly vanilla, non-private cluster I can see the same messages get logged when I omit that. I see it's defined in the template you provided above, but is it possible that it got removed after it was installed?

**Comment by mjnovice** [comment]
@nojnhuh you are correct. that error goes away after I specify an explicit `.spec.template.spec.version` on the MachinePool. 

However, I still see the following
```
I0627 23:21:18.658765       1 spec.go:427] managedclusters.Service.Parameters "msg"="found a diff between the desired spec and the existing managed cluster" "AzureManagedControlPlane"={"name":"ci-mj3","namespace":"sfd-ci-mj3"} "controller"="azuremanagedcontrolplane" "controllerGroup"="infrastructure.cluster.x-k8s.io" "controllerKind"="AzureManagedControlPlane" "difference"="  &containerservice.ManagedCluster{\n  \t... // 2 identical fields\n  \tExtendedLocation: nil,\n  \tIdentity:         nil,\n  \tManagedClusterProperties: &containerservice.ManagedClusterProperties{\n  \t\t... // 20 identical fields\n  \t\tAutoUpgradeProfile: nil,\n  \t\tAutoScalerProfile:  nil,\n  \t\tAPIServerAccessProfile: &containerservice.ManagedClusterAPIServerAccessProfile{\n- \t\t\tAuthorizedIPRanges:   &nil,\n+ \t\t\tAuthorizedIPRanges:   nil,\n  \t\t\tEnablePrivateCluster: nil,\n  \t\t\tPrivateDNSZone:       nil,\n  \t\t\t... // 2 identical fields\n  \t\t},\n  \t\tDiskEncryptionSetID: nil,\n  \t\tIdentityProfile:     nil,\n  \t\t... // 5 identical fields\n  \t},\n  \tTags:     nil,\n  \tLocation: nil,\n  \t... // 4 identical fields\n  }\n" "name"="ci-mj3" "namespace"="sfd-ci-mj3" "reconcileID"="bc24f73d-f142-4600-9eb3-86f8fc364edf" "x-ms-correlation-request-id"="ad75d889-ad36-450b-8cfd-6e768073995d"
```
and see redundant updates happening with Private cluster.

**Comment by nojnhuh** [comment]
Thanks for taking a look. Let's keep this issue open to track making the version field truly optional when backed by an AzureManagedMachinePool.

/retitle MachinePool version is optional but implicitly required by AzureManagedMachinePool

I think the AuthorizedIPRanges diff should be fixed by #3639.

**Comment by dtzar** [comment]
There has been a ton of code changes to this area in the past year.  I'm going to close this out, but happy to re-open if you're still experiencing this @mjnovice
